docs: blog post on parserOptions.project = true (#7368)
typescript-eslint/typescript-eslint: packages/website/blog/2023-08-02-parser-options-project-true.md

It converts raw TypeScript code into an ["AST" format](./2022-12-05-asts-and-typescript-eslint.md).
When [`parserOptions.project`](https://typescript-eslint.io/packages/parser#project) is specified, it additionally sets up TypeScript programs that can be used by [typed rules](https://typescript-eslint.io/developers/custom-rules#typed-rules).

Many projects today start with ESLint configs that look something like:

```js
module.exports = {
  // ...
  parserOptions: {
    project: './tsconfig.json',
    tsconfigRootDir: __dirname,
  },
  // ...
};
```

In larger repos, `parserOptions.project` often ends up being one of the three traditionally allowed forms:

- Path, such as `project: './tsconfig.json'`
- Glob pattern, such as `project: './packages/**/tsconfig.json'`
- Array of paths and/or glob patterns, such as `project: ['./packages/**/tsconfig.json', './separate-package/tsconfig.json']`

Explicitly indicating which TSConfig files are used for typed linting can be useful.
Developers like being given explicit control over their tooling.
However, we've seen a few issues arise from this approach:

> - Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)

Reviewer: ```suggestion
- Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)
```

- [Enhancement: Allow parserOptions.project to be (true | string)[]?](https://github.com/typescript-eslint/typescript-eslint/issues/7384)

If either of those two issues would benefit you, please 👍 react to them.
And if your project has a use case not yet mentioned in their comments, please post that use case.
We want to know what's important for users!

### Project Services

The downside of having users specify `parserOptions.project` at all is that `@typescript-eslint/parser` needs manual logic to create TypeScript Programs and associate them with linted files.
Manual Program creation logic comes with a few issues:

- Complex project setups can be difficult to get right.
  - For example, [typescript-eslint does not yet support Project References](https://github.com/typescript-eslint/typescript-eslint/issues/2094).
- The TypeScript compiler options used in the user's editor might differ from the compiler options in the TSConfigs they specified on disk.
- Files not included in created Programs can't be linted with type information, even though editors still typically surface type information when editing those files.

>   - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).

Reviewer: ```suggestion
  - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).
```

Many projects today start with ESLint configs that look something like:

```js
module.exports = {
  // ...
  parserOptions: {
    project: './tsconfig.json',
    tsconfigRootDir: __dirname,
  },
  // ...
};
```

In larger repos, `parserOptions.project` often ends up being one of the three traditionally allowed forms:

- Path, such as `project: './tsconfig.json'`
- Glob pattern, such as `project: './packages/**/tsconfig.json'`
- Array of paths and/or glob patterns, such as `project: ['./packages/**/tsconfig.json', './separate-package/tsconfig.json']`

Explicitly indicating which TSConfig files are used for typed linting can be useful.
Developers like being given explicit control over their tooling.
However, we've seen a few issues arise from this approach:

- Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)
- Needing to change a template ESLint config every time it's used for a different repository structure is a pain

> - Using different TSConfigs than what your editor users can result in different lint reports in the editor verses on the command-line

Author: Yeah if you run `eslint --config eslint.some.other.config.js` then you can give it totally different config values than what's in `eslint.config.js`.

For each file, `@typescript-eslint/parser` will check that file's directory, then the parent directory, and so on - until a `tsconfig.json` file is found.

:::tip
We recommend specifying [`tsconfigRootDir`](http://localhost:3000/packages/parser#tsconfigrootdir) in ESLint configs with a value set to the project's root directory (most commonly, `__dirname`).
That way, if you accidentally delete or rename the root `tsconfig.json` file, `@typescript-eslint/parser` won't search parent directories for higher `tsconfig.json` files.
:::

### Why Try `true`

If your project uses typed linting and manually specifies `tsconfig.json` files, we'd highly recommend trying out `parserOptions.project: true`.
We've seen it reduce lines of codes of ESLint configurations in many projects that have switched to it.
It's even occasionally even reduced time spent on typed linting by helping projects use a simpler set of TSConfigs. 🚀

In the long term, we're hoping to further improve the configuration and performance for typed linting (see _[Project Services](#project-services)_ below).

> Simplifying your configuration now will make it easier to onboard to our new options when they're available.

Author: I believe the current way is correct: one "onboards to" something.

### How It Works

When `@typescript-eslint/parser` is configured to generate type information, it attaches a backing TypeScript "Program" for to file it parses.
Those Programs provide type checking APIs used by lint rules.

> Each TSConfig file on disk is generally used to create exactly one Program, so files that refer to the same TSConfig file will reuse the same Program.

Reviewer: ```suggestion
Each TSConfig file on disk is generally used to create exactly one Program, and files included by the same TSConfig file will reuse the same Program.
```

Depending on how the ESLint config's `parserOptions.project` was specified, determining _which_ TSConfig file to use for each file can be different:

> - For a single string (e.g. `"tsconfig.json"`), then only one Program will be created, and all linted files will reuse it.

Reviewer: ```suggestion
- For a single path (e.g. `"tsconfig.json"`), only one Program will be created, and all linted files will reuse it.
```

> - For globs and/or arrays (e.g. `"./packages/*/tsconfig.json"`), then each linted file will reuse a Program based on the _first_ matched TSConfig file.

Reviewer: ```suggestion
- For globs and/or arrays (e.g. `"./packages/*/tsconfig.json"`), each linted file will use the Program created by the _first_ matched TSConfig file.
```

- Complex project setups can be difficult to get right.
  - For example, [typescript-eslint does not yet support Project References](https://github.com/typescript-eslint/typescript-eslint/issues/2094).
- The TypeScript compiler options used in the user's editor might differ from the compiler options in the TSConfigs they specified on disk.
- Files not included in created Programs can't be linted with type information, even though editors still typically surface type information when editing those files.
  - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).

We're working on an option to instead call the same TypeScript "Project Service" APIs that editors such as VS Code use to create Programs for us instead.
Project Services will automatically detect the TSConfig for each file (like `project: true`), and will also allow type information to be computed for JavaScript files without the `allowJs` compiler option (unlike `project: true`).

> We're hopeful this option will eventually become the standard way to enable typed linting.

Reviewer: ```suggestion
We hope this option will eventually become the standard way to enable typed linting.
```

However, because it's so new and untested, we're keeping it under the `EXPERIMENTAL_` prefix for at least all of the `6.X` versions.

See [Packages > Parser > `EXPERIMENTAL_useProjectService`](/packages/parser#EXPERIMENTAL_useProjectService) for more information.

## Supporting typescript-eslint

> If you enjoyed this blog post and/or or use typescript-eslint, please consider [supporting us on Open Collective](https://opencollective.com/typescript-eslint).

Reviewer: ```suggestion
If you enjoyed this blog post and/or use typescript-eslint, please consider [supporting us on Open Collective](https://opencollective.com/typescript-eslint).
```
